jetty-core/jetty-websocket/jetty-websocket-core-common/src/main/java/org/eclipse/jetty/websocket/core/messages/PartialStringMessageSink.java

```diff
@@ -42,12 +43,17 @@ public void accept(Frame frame, Callback callback)
             out.append(frame.getPayload());
             if (frame.isFin())
             {
-                methodHandle.invoke(out.toString(), true);
+                String complete = out.takeString(() -> new BadPayloadException("Invalid UTF-8"));
+                methodHandle.invoke(complete, true);
                 out = null;
             }
             else
             {
-                methodHandle.invoke(out.takePartialString(), false);
+                String partial = out.toString();
+                if (out.hasCodingErrors())
+                    throw new BadPayloadException("Invalid UTF-8");
+                out.partialReset();
+                methodHandle.invoke(partial, false);
```

Comment:
This API seems weird that we don't have the same style for taking full and partial strings. 

Can't we add a method `takePartialString` in same style as `takeString` and avoid having these external `hasCodingError()` and `partialReset()` calls every time it is used.

```
out.takePartialString(() -> new BadPayloadException("Invalid UTF-8"));
```

Reply:
@lachlan-roberts it is a little complicated with the naming.
`takeString` should really be called `takeCompleteString`, but while that name makes sense for some character codes, it is not so much for things like iso-8859-1.  But then there are other multi-character codes served by `CharSequenceBuilder` interface, so perhaps doing that rename makes sense.

We could then re-introduce a `takeString` API that takes the partial string and matches to `toString` semantic.

@joakime @lachlan-roberts  thoughts?

Reply:
I think its quite clear to have
- `takeString(Supplier<X>)`
- `takePartialString(Supplier<X>)`
- `toString()`

I'm okay with the rename of `takeString` -> `takeCompleteString`, but think it is unnecessary.
I think renaming `takePartialString` -> `takeString` is going to be very confusing.

Reply:
`takeString()` is a different API that `Utf8Appendable` implements via the `CharsetStringBuilder` interface, that name is already defined for other purposes.
It has to be a completed String.

Reply:
The fact that `toString()` is a partial String currently in this PR is actually tricky.  The overwhelming majority of uses of `toString()` are now bad and should be changed to `toCompleteString()`.

See commit f8b4884cfec8891d75ccd9aa642216cc1ab2ef3c here that fixes this bug.
(Submitted as a single commit to make it easier to revert)